packages/console/src/hooks/use-position.ts

```diff
@@ -0,0 +1,201 @@
+import { RefObject, useCallback, useEffect, useState } from 'react';
+
+export type VerticalAlignment = 'top' | 'bottom';
+
+export type HorizontalAlignment = 'start' | 'center' | 'end';
+
+type Offset = {
+  vertical: number;
+  horizontal: number;
+};
+
+type Props = {
+  verticalAlign: VerticalAlignment;
+  horizontalAlign: HorizontalAlignment;
+  offset: Offset;
+  anchorRef: RefObject<Element>;
+  overlayRef: RefObject<Element>;
+};
+
+type Position = {
+  top: number;
+  left: number;
+};
+
+// Leave space for box-shadow effect.
+const windowSafePadding = 12;
+
+const selectVerticalAlignment = ({
+  verticalAlign,
+  verticalTop,
+  verticalBottom,
+  overlayHeight,
+}: {
+  verticalAlign: VerticalAlignment;
+  verticalTop: number;
+  verticalBottom: number;
+  overlayHeight: number;
+}) => {
+  const minY = windowSafePadding;
+  const maxY = window.innerHeight - windowSafePadding;
+
+  const isTopAllowed = verticalTop >= minY;
+  const isBottomAllowed = verticalBottom + overlayHeight <= maxY;
+
+  if (verticalAlign === 'top') {
+    if (isTopAllowed) {
+      return 'top';
+    }
+
+    return isBottomAllowed ? 'bottom' : 'top';
+  }
+
+  // Mark: VerticalAlign === 'bottom'
+  if (isBottomAllowed) {
+    return 'bottom';
+  }
+
+  return isTopAllowed ? 'top' : 'bottom';
+};
+
+const selectHorizontalAlignment = ({
+  horizontalAlign,
+  horizontalStart,
+  horizontalCenter,
+  horizontalEnd,
+  overlayWidth,
+}: {
+  horizontalAlign: HorizontalAlignment;
+  horizontalStart: number;
+  horizontalCenter: number;
+  horizontalEnd: number;
+  overlayWidth: number;
+}) => {
+  const minX = windowSafePadding;
+  const maxX = window.innerWidth - windowSafePadding;
+
+  const isStartAllowed = horizontalStart + overlayWidth <= maxX;
+  const isCenterAllowed =
+    horizontalCenter - overlayWidth / 2 >= minX && horizontalCenter + overlayWidth / 2 <= maxX;
+  const isEndAllowed = horizontalEnd >= minX;
+
+  if (horizontalAlign === 'start') {
+    if (isStartAllowed) {
+      return 'start';
+    }
+
+    if (isEndAllowed) {
+      return 'end';
+    }
+
+    return isCenterAllowed ? 'center' : 'start';
+  }
+
+  if (horizontalAlign === 'end') {
+    if (isEndAllowed) {
+      return 'end';
+    }
+
+    if (isStartAllowed) {
+      return 'start';
+    }
+
+    return isCenterAllowed ? 'center' : 'end';
+  }
+
+  // Mark: horizontalAlign === 'center'
+  if (isCenterAllowed) {
+    return 'center';
+  }
+
+  if (isStartAllowed) {
+    return 'start';
+  }
+
+  if (isEndAllowed) {
+    return 'end';
+  }
+
+  return 'center';
```

Comment:
`return horizontalAlign;`

then line 91 and line 103 can be simplified

Reply:
updated